binderhub/registry.py

```diff
@@ -224,3 +227,61 @@ def get_image_manifest(self, image, tag):
                 raise
         else:
             return json.loads(resp.body.decode("utf-8"))
+
+
+class AWSElasticContainerRegistry(DockerRegistry):
+    aws_region = Unicode(
+        config=True,
+        help="""
+        AWS region for ECR service
+        """,
+    )
+
+    ecr_client = Any()
+
+    @default("ecr_client")
+    def _get_ecr_client(self):
+        return boto3.client("ecr", region_name=self.aws_region)
+
+    # TODO: cache auth if not expired - authorizationData[i]["expiresAt"]
+    def _get_ecr_auth(self):
+        auths = self.ecr_client.get_authorization_token()["authorizationData"]
+        auth = next(x for x in auths if x["proxyEndpoint"] == self.url)
+        self._patch_docker_config_secret(auth)
```

Comment:
Could you explain a bit why we need to modify the push secret? All other registries work with a static push secret.

What happens when the build container is trying to push an image and the push secret is wrong/changed?

Reply:
With AWS ECR, an IAM principal is used to generate an auth token that is valid for 12 hours - source: https://docs.aws.amazon.com/AmazonECR/latest/userguide/Registries.html

So what we're doing here is using the AWS boto3 client to get/generate an auth token and update the push secret with it, so the builder picks up a fresh, valid token in the case where it doesn't exist already. There is admittedly an edge case here if the builder takes longer than 12 hours to build, but I believe the likelihood in practice is acceptably low.